pkg/admission/zone_tracker.go

```diff
@@ -0,0 +1,231 @@
+package admission
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/rollout-operator/pkg/config"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/objstore/providers/azure"
+	"github.com/thanos-io/objstore/providers/gcs"
+	"github.com/thanos-io/objstore/providers/s3"
+	"gopkg.in/yaml.v2"
+	appsv1 "k8s.io/api/apps/v1"
+)
+
+type zoneTracker struct {
+	mu    sync.Mutex
+	zones map[string]string
+	bkt   objstore.Bucket
+	key   string
+}
+
+// loadZones loads the zone file from object storage and populates the zone tracker.
+// If the zone file does not exist, create it and upload it to the bucket then try to get it again.
+func (zt *zoneTracker) loadZones(ctx context.Context, stsList *appsv1.StatefulSetList) error {
+	r, err := zt.bkt.Get(ctx, zt.key)
+	if err != nil {
+		if zt.bkt.IsObjNotFoundErr(err) {
+			// Create the zone file if it doesn't exist and populate with the current time for each zone, try to get it again
+			err = zt.createInitialZones(ctx, stsList)
+			if err != nil {
+				return err
+			}
+			r, err = zt.bkt.Get(ctx, zt.key)
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+	defer r.Close()
+
+	return json.NewDecoder(r).Decode(&zt.zones)
+}
+
+// saveZones encodes the zone tracker to json and uploads the zone file to the bucket.
+func (zt *zoneTracker) saveZones(ctx context.Context) error {
```

Comment:
Maybe add a doc that callers should lock zt.mu before calling

Reply:
Adding.